backend/cfg/cfg_dataflow.ml

```diff
@@ -1,151 +1,320 @@
+        | c -> c
+    end
 
-  type map = domain Label.Tbl.t
+    module WorkSet = Set.Make (WorkSetElement)
 
-  module WorkSetElement = struct
     type t =
-      { label : Label.t;
-        value : domain
+      { priorities : int Label.Tbl.t;
+        mutable work_set : WorkSet.t
       }
 
-    let compare { label = left_label; value = left_value }
-        { label = right_label; value = right_value } =
-      match Label.compare left_label right_label with
-      | 0 -> D.compare left_value right_value
-      | res -> res
+    type element = Label.t
+
+    let create ~priorities = { priorities; work_set = WorkSet.empty }
+
+    let add t label =
+      let priority = Label.Tbl.find t.priorities label in
+      t.work_set <- WorkSet.add { label; priority } t.work_set
+
+    let is_empty t = WorkSet.is_empty t.work_set
+
+    let choose t = WorkSet.max_elt t.work_set
+
+    let remove_and_return t =
+      let element = choose t in
+      t.work_set <- WorkSet.remove element t.work_set;
+      element.label
   end
 
-  module WorkSet = Set.Make (WorkSetElement)
+  type work_state =
+    { cfg : Cfg.t;
+      mutable queue : WorkSet.t;
+      map_block : Transfer_domain.t Label.Tbl.t;
+      map_instr : D.instr_domain Instr.Tbl.t option
+    }
+
+  type instr_domain = D.instr_domain
 
-  let transfer_block : domain -> Cfg.basic_block -> transfer =
-   fun value block ->
-    let transfer f (acc_normal, acc_exceptional) instr =
-      let { T.normal; exceptional } = f acc_normal instr in
-      normal, D.join exceptional acc_exceptional
-    in
-    let normal, exceptional =
-      transfer T.terminator
-        (ListLabels.fold_left block.body ~init:(value, value)
-           ~f:(transfer T.basic))
-        block.terminator
+  let compute_priorities (cfg : Cfg.t) =
+    let priorities = Label.Tbl.create (Label.Tbl.length cfg.blocks) in
+    Cfg.iter_blocks cfg ~f:(fun label _block ->
+        (* Explicitly set each block priority to (-label). With that blocks will
+           be computed in order of increasing labels and that's the same as the
+           previous implicit order. *)
+        Label.Tbl.add priorities label (-label));
+    assert (Label.Tbl.length priorities = Label.Tbl.length cfg.blocks);
+    priorities
+
+  let update_instr : work_state -> int -> instr_domain -> unit =
+   fun t instr_id value ->
+    match t.map_instr with
+    | None -> ()
+    | Some map_instr -> Instr.Tbl.replace map_instr instr_id value
+
+  let create :
+      Cfg.t ->
+      init:(Cfg.basic_block -> Transfer_domain.t option) ->
+      store_instr:bool ->
+      work_state =
+   fun cfg ~init ~store_instr ->
+    let priorities = compute_priorities cfg in
+    let queue = WorkSet.create ~priorities in
+    let map_block = Label.Tbl.create (Label.Tbl.length cfg.Cfg.blocks) in
+    let map_instr =
+      if store_instr
+      then
+        let map_instr =
+          Instr.Tbl.create (Label.Tbl.length cfg.Cfg.blocks * 16)
```

Comment:
(I think we have lost a CR-something
about that constant - or are we positive
it makes sense to simply use `16`?)

Reply:
Added it back.